### PR TITLE
instantsend: Introduce KnownInstantSendLock, optimize ProcessInstantSendLock a bit

### PR DESCRIPTION
**Pull request by UdjinM6**
Please see individual commits.

Extracted from #3970, ~based on #3983 atm~

### CODE REVIEW

File: src/llmq/quorums_instantsend.cpp [hunk 1]
```diff
@@ -904,10 +880,33 @@ void CInstantSendManager::ProcessInstantSendLock(NodeId from, const uint256& has
         creatingInstantSendLocks.erase(islock->GetRequestId());
         txToCreatingInstantSendLocks.erase(islock->txid);
 
-        CInstantSendLockPtr otherIsLock;
-        if (db.GetInstantSendLockByHash(hash)) {
+        if (pendingInstantSendLocks.count(hash) || db.KnownInstantSendLock(hash)) {
```
Comment (PastaPastaPasta) [inline src/llmq/quorums_instantsend.cpp:883]: why'd you add the `pendingInstantSendLocks.count(hash)` here?

File: src/llmq/quorums_instantsend.cpp [hunk 2]
```diff
@@ -904,10 +880,33 @@ void CInstantSendManager::ProcessInstantSendLock(NodeId from, const uint256& has
...
+    // we ignore failure here as we must be able to propagate the lock even if we don't have the TX locally
+    if (GetTransaction(islock->txid, tx, Params().GetConsensus(), hashBlock) && !hashBlock.IsNull()) {
+        {
+            LOCK(cs_main);
+            pindexMined = LookupBlockIndex(hashBlock);
+        }
+
+        // Let's see if the TX that was locked by this islock is already mined in a ChainLocked block. If yes,
+        // we can simply ignore the islock, as the ChainLock implies locking of all TXs in that chain
+        if (llmq::chainLocksHandler->HasChainLock(pindexMined->nHeight, pindexMined->GetBlockHash())) {
```
Comment (PastaPastaPasta) [inline src/llmq/quorums_instantsend.cpp:900]: We should probably assert that pindexMined isn't null, though that's arguable outside of this PRs scope

File: src/llmq/quorums_instantsend.cpp [hunk 3]
```diff
@@ -904,10 +880,33 @@ void CInstantSendManager::ProcessInstantSendLock(NodeId from, const uint256& has
...
+            LogPrint(BCLog::INSTANTSEND, "CInstantSendManager::%s -- txlock=%s, islock=%s: dropping islock as it already got a ChainLock in block %s, peer=%d\n", __func__,
+                     islock->txid.ToString(), hash.ToString(), hashBlock.ToString(), from);
+            return;
+        }
+    }
+
+    {
+        LOCK(cs);
+        CInstantSendLockPtr otherIsLock;
         otherIsLock = db.GetInstantSendLockByTxid(islock->txid);
```
Comment (xdustinface) [inline src/llmq/quorums_instantsend.cpp:910]:
```suggestion
        CInstantSendLockPtr otherIsLock = db.GetInstantSendLockByTxid(islock->txid);
```

**Review by xdustinface — CHANGES_REQUESTED** [review]
Looks good, but see comment below for one nit.

**Review by xdustinface — APPROVED** [review]
utACK

File: src/llmq/quorums_instantsend.h [hunk 4]
```diff
@@ -60,7 +60,8 @@ class CInstantSendDb
     static void WriteInstantSendLockArchived(CDBBatch& batch, const uint256& hash, int nHeight);
     std::unordered_map<uint256, CInstantSendLockPtr> RemoveConfirmedInstantSendLocks(int nUntilHeight);
     void RemoveArchivedInstantSendLocks(int nUntilHeight);
-    bool HasArchivedInstantSendLock(const uint256& islockHash) const;
+
```
Comment (PastaPastaPasta) [inline src/llmq/quorums_instantsend.h:63]: extra line?
```suggestion
```

**Review by xdustinface — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Pull request merged**